pkg/kvstore/redis.go

```diff
@@ -162,13 +169,45 @@ func (s *redisKvStore) PutBatch(ctx context.Context, values interface{}) error {
 		return fmt.Errorf("could not convert values from %T to map[interface{}]interface{}", values)
 	}
 
+	chunkSize := s.settings.BatchSize
+	pairs := make([]interface{}, 0, 2*chunkSize)
 	for k, v := range mii {
-		if err = s.Put(ctx, k, v); err != nil {
-			return fmt.Errorf("failed to write batch to redis: %w", err)
+		key, value, err := s.marshalKeyValue(k, v)
+		if err != nil {
+			return fmt.Errorf("PutBatch could not marshal key/value: %w", err)
+		}
+		pairs = append(pairs, key, value)
+
+		if len(pairs) >= 2*chunkSize {
+			err = s.flushChunk(ctx, pairs)
+			if err != nil {
+				return fmt.Errorf("failed to write batch to redis: %w", err)
+			}
+			pairs = make([]interface{}, 0, chunkSize)
 		}
 	}
 
-	return nil
+	return s.flushChunk(ctx, pairs)
+}
+
+func (s *redisKvStore) flushChunk(ctx context.Context, pairs []interface{}) error {
+	if len(pairs) < 1 {
+		return nil
+	}
+
+	pipe := s.client.Pipeline().TxPipeline()
+
+	pipe.MSet(ctx, pairs)
+	// setting ttl
+	for i := 0; i < len(pairs); i += 2 {
+		keyStr, ok := pairs[i].(string)
+		if !ok {
+			return fmt.Errorf("setting ttl, failed to cast key to string: %v", pairs[i])
+		}
+		pipe.Expire(ctx, keyStr, s.settings.Ttl)
```

Comment:
Do we have a ttl in every case? Or is the ttl optional?  
So we need a call for every single pair? I guess no way to do this in batch as well?

Reply:
don't know of a batch TTL command,
pipelining was the next best thing.